### PR TITLE
Update sequelize.sync() promise to return self.  See #3530

### PR DESCRIPTION
**Pull request by MichaelJCole**
Update Sequalize.sync() to return a promise resolving to itself.  Tested locally with sqlite, mysql, postgress.  Integrations failed on my box for both patched and unpatched versions.  I'm guessing you have a test env already setup.

Thanks!

Mike

### CODE REVIEW

File: lib/sequelize.js [hunk 1]
```diff
@@ -893,9 +893,12 @@ module.exports = (function() {
         }
       });
 
+      // Iterate daos and sync.  Collect results into array
       return Promise.reduce(daos, function(total, dao) {
-        return dao.sync(options);
-      }, null);
+        return total.push(dao.sync(options));
```
Comment (mickhansen) [inline lib/sequelize.js]: `push` returns the target so the next iteration of reduce will fail.
You need to use something like:

```js
return dao.sync(options).then(function (result) {
  return total.concat([result]);
});
```

Comment (MichaelJCole) [inline lib/sequelize.js]: Hey Mick, thanks.  Fixed this a different way, but I like your syntax better.  Just a sec and I'll commit.

File: lib/sequelize.js [hunk 2]
```diff
@@ -893,9 +893,13 @@ module.exports = (function() {
...
       });
 
+      // Iterate daos and sync.  Collect results into array
       return Promise.reduce(daos, function(total, dao) {
-        return dao.sync(options);
-      }, null);
+        total.push(dao.sync(options));
+        return total;
+      }, [])
+      // Array is ok, but I really want 'this' as DI container for models
```
Comment (mickhansen) [inline lib/sequelize.js]: Remove the comment.

File: lib/sequelize.js [hunk 3]
```diff
@@ -893,9 +893,13 @@ module.exports = (function() {
         }
       });
 
+      // Iterate daos and sync.  Collect results into array
       return Promise.reduce(daos, function(total, dao) {
-        return dao.sync(options);
-      }, null);
+        total.push(dao.sync(options));
```
Comment (mickhansen) [inline lib/sequelize.js]: I'm not sure you can push the promise, i'm not sure it works that way.

**Comment by MichaelJCole** [comment]
Hey, I've had a bit of a learning curve on getting the tests to run.  I've got the tests running now.  Stand by for a fix.

**Comment by BridgeAR** [comment]
Use:

```javascript
return Promise.resolve (daos).each (function( dao) {
    return dao.sync(options);
});
```

Promise reduce was used here to use a promise chain. This should be done using each instead. It'll resolve to the original array.

**Comment by mickhansen** [comment]
@BridgeAR is each serial?

**Comment by mickhansen** [comment]
Ah yes, so that is, that is a cleaner solution then :)

**Comment by MichaelJCole** [comment]
Ok, sorry for all the message traffic.  I took out the extra reduce() code and just .return(this).  Which is what you said in your first comment Mick :-)

Travis is failing on the very last config iojs w/ mssql.  I don't really understand the build system, but it appears to be having [trouble logging into the DB](https://travis-ci.org/sequelize/sequelize/jobs/58402655#L301).

Is this an intermittent error?

**Comment by BridgeAR** [comment]
This should be closed.

**Pull request closed without merging**